gobblin-config-management/gobblin-config-core/src/main/java/gobblin/config/store/hdfs/SimpleHDFSStoreMetadata.java

```diff
@@ -0,0 +1,133 @@
+import java.io.InputStreamReader;
+
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.base.Charsets;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+import com.typesafe.config.ConfigRenderOptions;
+import com.typesafe.config.ConfigValueFactory;
+
+
+/**
+ * A metadata accessor for {@link SimpleHDFSConfigStore}. Every {@link SimpleHDFSConfigStore} will have a file named
+ * {@link #CONFIG_STORE_METADATA_FILENAME} that contains store metadata.
+ * For instance the current active version of the store is stored at {@link #CONFIG_STORE_METADATA_CURRENT_VERSION_KEY}.
+ * This class helps reading and writing store metadata in {@link #CONFIG_STORE_METADATA_FILENAME}
+ */
+public class SimpleHDFSStoreMetadata {
+
+  private static final String CONFIG_STORE_METADATA_FILENAME = "store-metadata.conf";
+  private static final String CONFIG_STORE_METADATA_CURRENT_VERSION_KEY = "config.hdfs.store.version.current";
+
+  private final FileSystem fs;
+  private final Path storeMetadataFilePath;
+
+  /**
+   * Create a new {@link SimpleHDFSStoreMetadata} to read and write store metadata
+   *
+   * @param fs where metadata is stored
+   * @param configStoreDir path to {@link SimpleHDFSConfigStore#CONFIG_STORE_NAME}
+   */
+  SimpleHDFSStoreMetadata(final FileSystem fs, final Path configStoreDir) {
+    this.storeMetadataFilePath = new Path(configStoreDir, CONFIG_STORE_METADATA_FILENAME);
+    this.fs = fs;
+  }
+
+  /**
+   * Writes the <code>config</code> to {@link #storeMetadataFilePath}. Creates a backup file at
+   * <code>storeMetadataFilePath + ".bkp"</code> to recover old metadata in case of unexpected deployment failures
+   *
+   * @param config to be serialized
+   */
+  void writeMetadata(Config config) throws IOException {
+
+    Path storeMetadataFileBkpPath = new Path(this.storeMetadataFilePath.getParent(), this.storeMetadataFilePath.getName() + ".bkp");
+
+    try {
+
+      // Delete old backup file if exists
+      if (this.fs.exists(storeMetadataFileBkpPath)) {
+        if (!this.fs.delete(storeMetadataFileBkpPath, true)) {
+          throw new IOException("Failed to delete " + storeMetadataFileBkpPath);
+        }
+      }
+
+      // Rename current storeMetadataFile to backup
+      if (this.fs.exists(this.storeMetadataFilePath)) {
+        if (!this.fs.rename(this.storeMetadataFilePath, storeMetadataFileBkpPath)) {
+          throw new IOException(String.format("Failed to create backup file for %s at %s", this.storeMetadataFilePath, storeMetadataFileBkpPath));
+        }
+      }
+
+      // Write new storeMetadataFile
+      try (FSDataOutputStream outputStream = this.fs.create(this.storeMetadataFilePath);) {
+        outputStream.write(config.root().render(ConfigRenderOptions.concise()).getBytes(Charsets.UTF_8));
+      }
+    } finally {
+      this.fs.delete(storeMetadataFileBkpPath, true);
```

Comment:
do we want to try and restore the back up file if there is a problem creating the new file?